### PR TITLE
Rewrite ShardedTensor.gather to use dist.gather instead of gather_object

### LINKED ISSUE #73935: [shard] use dist.gather instead of dist.gather_object
### 🚀 The feature, motivation and pitch

See context https://github.com/pytorch/pytorch/pull/71624

Coping some here: 

Now we have `dist.gather` support on NCCL, but `ShardedTensor.gather` couldn't use it for some reasons: 

1. We can use gather_object for metadatas, but we couldn't simply use gather for the local shard tensors, mainly because local_shards() is a list of tensor, but input of gather is a single tensor not a list of tensor, so gather won't work here
2. We can try using torch.cat locally on each rank to form a single tensor before the gather collective, but we need to split them afterwards as they might not be adjacent to each other (local shards on this rank might contain two tensors that's far away in the global position logically). So we might need to insert additional insertion point in the first gather_object to split the combined tensors from the second gather call. This is pretty tricky from my understanding.

### Alternatives

We are right now using `dist.gather_object`, should be slightly better than `dist.allgather_object` but ideally we want to avoid pickling

### Additional context

_No response_

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### PR DESCRIPTION
**Pull request by kumpera**
gather_object is problematic when used with Tensors as they can unpickle on the wrong
device and lead to deadlocks or spurious failures.

This change introduces a RPC workaround for EFA when initing TensorPipe until
they properly address it.

Fixes #73935

### CODE REVIEW

File: test/distributed/_shard/sharded_tensor/test_sharded_tensor.py [hunk 1]
```diff
@@ -918,7 +919,7 @@ def test_invalid_pg_rpc_ranks(self):
         self.init_pg()
 
         # Init RPC with different ranks.
-        rpc_backend_options = rpc.TensorPipeRpcBackendOptions()
+        rpc_backend_options = rpc.TensorPipeRpcBackendOptions(_transports=tp_transports())
```
Comment (pritamdamania87) [inline test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:922]: Do the EFA workarounds have anything to do with `ShardedTensor.gather`? If not, can we move these changes to a separate PR?

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 2]
```diff
@@ -273,46 +278,85 @@ def gather(
                 Must to be provided ONLY on ``dst`` rank.
                 Default: ``None``
         """
+        def shard_size(shard_md):
+            res = 1
+            for s in shard_md.shard_sizes:
+                res *= s
+            return res
```
Comment (pritamdamania87) [inline torch/distributed/_shard/sharded_tensor/api.py:285]: Can't we just do `math.prod(shard_md.shard_sizes)`?

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 3]
```diff
@@ -273,46 +278,85 @@ def gather(
...
+            gather_list = None
+
+        with torch.no_grad():
+            # This is an aproximation for what device to use.
+            # The lack of policy enforcement around is pretty bad
+            device: Union[torch.device, str]
+            if len(local_shards) > 0:
+                device = local_shards[0].tensor.device
+            else:
+                device = f"cuda:{torch.cuda.current_device()}"
```
Comment (pritamdamania87) [inline torch/distributed/_shard/sharded_tensor/api.py:322]: Why don't we always use the current device? We do expect the user to set this appropriately as documented here: https://github.com/pytorch/pytorch/blob/master/torch/distributed/_shard/sharded_tensor/api.py#L109

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 4]
```diff
@@ -273,46 +278,85 @@ def gather(
...
+            data = torch.empty(max_rank_size, device=device)
+            for shard in local_shards:
+                for placement in local_shards_placement:
```
Comment (pritamdamania87) [inline torch/distributed/_shard/sharded_tensor/api.py:325]: Why do we need `local_shards_placement`? Doesn't `local_shards` have everything you need?

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 5]
```diff
@@ -273,46 +278,85 @@ def gather(
...
-                            dim,
-                            metadata.shard_offsets[dim],
-                            metadata.shard_sizes[dim],
-                        )
-
-                    out_narrow_view.copy_(tensor)
+        if rank != dst:
+            return
+        if out is None:
+            raise ValueError("`out` Tensor must be provided on dst rank!")
```
Comment (pritamdamania87) [inline torch/distributed/_shard/sharded_tensor/api.py:340]: Probably do this validation right at the top of the function, not after doing expensive collectives.

**Review by pritamdamania87 — COMMENTED** [review]
Thanks for fixing this!

File: torch/distributed/_shard/sharded_tensor/api.py (hunk 3 above)
Comment (rohan-varma) [inline torch/distributed/_shard/sharded_tensor/api.py:322]: nit: I don't think the device selection needs to be under no_grad context manager?

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 6]
```diff
@@ -35,6 +35,7 @@
     load_tests,
     sandcastle_skip_if,
     get_cycles_per_ms,
+    tp_transports,
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/rpc/rpc_test.py:38]: Can we separate out the RPC changes to a different PR?

File: test/distributed/_shard/sharded_tensor/test_sharded_tensor.py (hunk 1 above)
Comment (kumpera) [reply test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:922]: Sure, it's here because I have a hard time working on it on AWS otherwise.

File: torch/distributed/_shard/sharded_tensor/api.py (hunk 3 above)
Comment (kumpera) [reply torch/distributed/_shard/sharded_tensor/api.py:322]: Always using CUDA is wrong if this is a CPU ShardedTensor.

What we want is to alloc on `self._process_group` device which should be `cuda.current_device` if we're using NCCL or cpu with GLOO.

What I'm thinking here, and I need to add a test for, is how to handle `sharded_tensor.cpu().gather()` when there's one rank without data.

File: torch/distributed/_shard/sharded_tensor/api.py (hunk 4 above)
Comment (kumpera) [reply torch/distributed/_shard/sharded_tensor/api.py:325]: This is a badly named variable, it tracks each shard offset in the exchange tensor we're building.

File: torch/distributed/_shard/sharded_tensor/api.py (hunk 5 above)
Comment (kumpera) [reply torch/distributed/_shard/sharded_tensor/api.py:340]: This check is already done by _validate_output_tensor_for_gather but mypy can't see it. 
Replacing it with a cast and a comment on why.

File: torch/distributed/_shard/sharded_tensor/api.py (hunk 3 above)
Comment (kumpera) [reply torch/distributed/_shard/sharded_tensor/api.py:322]: good point

Comment (kumpera) [reply torch/distributed/_shard/sharded_tensor/api.py:322]: I replaced this with adding a method `_get_preferred_device` to ST which returns cuda:current_device for NCCL or CPU otherwise.

File: torch/testing/_internal/distributed/rpc/rpc_test.py (hunk 6 above)
Comment (kumpera) [reply torch/testing/_internal/distributed/rpc/rpc_test.py]: done

File: torch/distributed/_shard/sharded_tensor/api.py (hunk 2 above)
Comment (kumpera) [reply torch/distributed/_shard/sharded_tensor/api.py]: TIL

File: test/distributed/_shard/sharded_tensor/test_sharded_tensor.py (hunk 1 above)
Comment (pritamdamania87) [reply test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:922]: Looks like we still have RPC changes here, shouldn't this go in https://github.com/pytorch/pytorch/pull/77363?

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 7]
```diff
@@ -273,46 +288,76 @@ def gather(
...
+        if rank == dst:
+            assert out is not None
+            gather_list = [torch.empty((max_rank_size,), device=out.device) for _ in range(world_size)]
+        else:
+            gather_list = None
+
+        with torch.no_grad():
+            data = torch.empty(max_rank_size, device=self._get_preferred_device())
+            for shard in local_shards:
+                for placement in local_shards_offset:
```
Comment (pritamdamania87) [inline torch/distributed/_shard/sharded_tensor/api.py:326]: nit: Might be cleaner if `local_shards_offsets` is a dict keyed on `ShardMetadata`, since `ShardMetadata` is a dataclass it probably already has `__eq__` and `__hash__` to enable this, if not maybe we can add it?

**Review by pritamdamania87 — APPROVED** [review]

File: torch/distributed/_shard/sharded_tensor/api.py (hunk 7 above)
Comment (kumpera) [reply torch/distributed/_shard/sharded_tensor/api.py]: I had to make ShardMetadata and _remote_device hashable.
Which those changes, the `gather` code became significantly simpler.
Good catch!

Comment (kumpera) [reply torch/distributed/_shard/sharded_tensor/api.py]: I made that change, had to make both ShardMetadata and _remote_device  hashable.
The code turned out to be significantly clearer.
